packages/router/src/navigation.tsx

```diff
@@ -0,0 +1,208 @@
+import type { ReactNode } from 'react'
+import { useCallback, useContext, useEffect, useMemo, useState } from 'react'
+
+import type { Location } from './location'
+import { useLocation } from './location'
+import { createNamedContext } from './util'
+
+export enum NavigationMethod {
+  PUSH = 'PUSH',
+  REPLACE = 'REPLACE',
+  POP = 'POP',
+  UNLOAD = 'UNLOAD',
+}
+
+export type NavigationAction = {
+  data: any
+  unused: string
+  url?: string | URL | null
+}
+
+export type NavigationBlockerFn = (
+  from: Location,
+  to: Location,
+  method: NavigationMethod,
+  confirm?: () => void
+) => boolean
+
+type NavigationBlocker = {
+  id: string
+  check: NavigationBlockerFn
+}
+
+type Navigation = {
+  from: Location
+  to: Location
+}
+
+export type NavigationState = {
+  previousLocation: Location
+  currentLocation: Location
+}
+
+export type NavigationContextState = {
+  navigation: Navigation
+  register: (blocker: NavigationBlockerFn) => string
+  unregister: (blockerId: string) => void
+}
+
+const NavigationContext =
+  createNamedContext<NavigationContextState>('NavigationState')
+
+type NavigationProviderProps = {
+  children: ReactNode
+}
+
+const NavigationProvider = ({ children }: NavigationProviderProps) => {
+  /**
+   * Blockers are functions that are called before a navigation action is performed
+   */
+  const [blockers, setBlockers] = useState<NavigationBlocker[]>([])
+
+  /**
+   * Navigation state is used to keep track of the current and previous location data
+   */
+  const location: Location = useLocation()
+  const [navigation, setNavigation] = useState({
+    from: location,
+    to: location,
+  })
+
+  useEffect(() => {
+    setNavigation((prev) => {
+      return {
+        from: prev.to,
+        to: location,
+      }
+    })
+  }, [location])
+
+  /**
+   * Register the interceptors and override the browser history methods
+   */
+  useEffect(() => {
+    const originalPushState = globalThis.history.pushState.bind(
+      globalThis.history
+    )
+    const originalReplaceState = globalThis.history.replaceState.bind(
+      globalThis.history
+    )
+
+    globalThis.history.pushState = function pushState(
+      data: any,
+      unused: string,
+      url?: string | URL | null
+    ) {
+      const currentLocation = new URL(globalThis.location.href)
+      const newLocation = new URL(globalThis.location.origin + url)
+
+      if (
+        currentLocation.pathname === newLocation.pathname &&
+        currentLocation.search === newLocation.search &&
+        currentLocation.hash === newLocation.hash
+      ) {
+        return
+      }
+
+      const blocked = blockers.some(({ check }) =>
+        check(currentLocation, newLocation, NavigationMethod.PUSH, () => {
+          originalPushState(data, unused, url)
+        })
+      )
+
+      if (!blocked) {
+        originalPushState(data, unused, url)
+      }
```

Comment:
pushState isn't invoke when a blocker returned true... but the URL changes to the new location anyway. (WHY??)